llvm/lib/Transforms/InstCombine/InstructionCombining.cpp

```diff
@@ -4842,8 +4847,11 @@ Instruction *InstCombinerImpl::visitFreeze(FreezeInst &I) {
 
   Constant *C;
   if (match(Op0, m_Constant(C)) && C->containsUndefOrPoisonElement()) {
```

Comment:
Can you add some vector tests?
```
define <2 x i8> @src(i1 %cond, <2 x i8> noundef %y) {
  %undef = freeze <2 x i8> <i8 0, i8 poison>
  %sel = select i1 %cond, <2 x i8> %y, <2 x i8> %undef
  ret <2 x i8> %sel
}

define <2 x i8> @tgt(i1 %cond, <2 x i8> noundef %y) {
  ret <2 x i8> %y
}
```
It is a miscompilation: https://alive2.llvm.org/ce/z/8Q3XYH

Reply:
BTW, as it is not suitable to handle this fold under `getUndefReplacement`, can we move the logic into `InstCombinerImpl::visitSelect`?
```
Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
...
    if (match(TrueVal, m_Freeze(m_Poison())) &&
        isGuaranteedNotToBeUndefOrPoison(FalseVal, &AC, &I, &DT))
      return replaceInstUsesWith(FalseVal);
    if (match(FalseVal, m_Freeze(m_Poison())) &&
        isGuaranteedNotToBeUndefOrPoison(TrueVal, &AC, &I, &DT))
      return replaceInstUsesWith(TrueVal);
```

Reply:
Thanks for pointing out the miscompilation!

The reason I did not place this logic into `InstCombinerImpl::visitSelect` was the canonical form produced by `InstCombinerImpl::visitFreeze` occurs first and thus ruins the match on freeze poison when visitation is performed on the select instruction.

I experimented a bit and derived a two match expression in the lambda function.

Added basic vector tests.

Reply:
> the canonical form produced by InstCombinerImpl::visitFreeze occurs first and thus ruins the match on freeze poison when visitation is performed on the select instruction.

Should we remove these canonicalizations first?